### ISSUE TITLE
ActiveSupport::JSON.decode Fails Decoding ISO8601 Date/Time String

### ISSUE THREAD

**Issue by trcarden** [rails/rails#22171]
When using activesupport to parse JSON a seemingly valid datetime string (``"2015-02-14T02:21:34"`) seems to not get parsed into a valid datetime.

Full example:
```
$ rails c
Loading development environment (Rails 4.2.4)
2.2.3 :001 > ActiveSupport.parse_json_times = True
2.2.3 :002 > ActiveSupport::JSON.decode('{"a":"2015-02-14T02:21:34"}')
 => {"a"=>"2015-02-14T02:21:34"}  # Incorrectly a string

2.2.3 :003 > ActiveSupport::JSON.decode('{"a":"2015-02-14T02:21:34Z"}')
 => {"a"=>Sat, 14 Feb 2015 02:21:34 +0000}  # Correctly a datetime
```

I believe the string `"2015-02-14T02:21:34"` is a valid iso8601 datetime string (it just doesn't have a timezone indicator). 

If you trace the source of this it goes to `activesupport/lib/activesupport/decoding.rb` and the regular expression

```
DATE_REGEX = /^(?:\d{4}-\d{2}-\d{2}|\d{4}-\d{1,2}-\d{1,2}[T \t]+\d{1,2}:\d{2}:\d{2}(\.[0-9]*)?(([ \t]*)Z|[-+]\d{2}?(:\d{2})?))$/
```

Should probably be
```
DATE_REGEX = /^(?:\d{4}-\d{2}-\d{2}|\d{4}-\d{1,2}-\d{1,2}[T \t]+\d{1,2}:\d{2}:\d{2}(\.[0-9]*)?(([ \t]*)Z|[-+]\d{2}?(:\d{2})?)?)$/
```
Note the `?` as the 4rd to last character in the expression

**Comment by chancancode** [comment]
Seems good! Do you mind sending a PR and adding a test?

**Comment by al2o3cr** [comment]
The standards are fairly unclear on this - ISO8601 (according to Wikipedia, anyways) specifies that datetimes without a TZ indicator are to be regarded as "local" time. Note that fixing the regex as described above DOES NOT fix this; `DateTime.parse` believes strings without TZ information to be in UTC.

[RFC3339](https://tools.ietf.org/html/rfc3339#page-8), on the other hand, explicitly requires either the `T`+offset form or the `Z`.

**Comment by chancancode** [comment]
@trcarden where did the string come from and what did you expect the parser to do re local vs UTC time?

**Comment by gaurish** [comment]
As per the standard, if there NO timezone it is treated as local time. this is the behaviour which is correctly implemented by `Time.iso8601` but `DateTime.iso8601`

```ruby
2.2.3 :007 > Time.iso8601 '2015-02-14T02:21:34'
2015-02-14 02:21:34 +0530 #Correct as per ISO Spec
2.2.3 :008 > DateTime.iso8601 '2015-02-14T02:21:34'
Sat, 14 Feb 2015 02:21:34 +0000 # InCorrect
```

so I think we should parse this & treat it as local time as [per ISO spec info on wikipedia](https://en.wikipedia.org/wiki/ISO_8601#Time_zone_designators)

**Comment by trcarden** [comment]
@chancancode that string comes from the official ISO8601 spec (you have to buy a copy to see it but there seems to be a older version available here http://www.oamk.fi/~karil/iso_standard_8601_for_dates_and_time.pdf which should get the point across). Section 5.4 titled `Combinations of date and time of the day representations` says a datetime timestamp without the offset or `Z|z` character to be interpreted as local time.

I got into the problem because python implements datetime.isoformat this way and outputs strings with that format. So either all of python is wrong (wouldn't be the first time) when dealing with locale datetime iso8601 strings or there is a bug in this parser.

**Comment by arnvald** [comment]
As @gaurish mentioned, the `DateTime` and `Time` behaviour is inconsistent. I'm going to submit this to https://bugs.ruby-lang.org and see if it will be fixed in the next version of Ruby (or maybe for some reason it's on purpose?)

@chancancode what to do about it in Rails? Should we fix the regexp so that it will work with or without `Z` even though behaviour won't be fully correct according to spec?
@trcarden if you're busy then I can add tests and open PR, just let me know.

**Comment by chancancode** [comment]
@pixeltrix @fxn do you have any thoughts?

Personally, I think it makes sense to fix the regexp to let it through – Python interop seems particularly important. The JSON spec doesn't say anything about dates so it's not really a standards concern.

However, in the current state, it seems like the only thing this will do is exposing users to a Ruby "bug". I'm not too sure what to do about that.

**Comment by pixeltrix** [comment]
I think my preferred preference would be to adjust the regexp and change the code to use `Time.iso8601` instead of `DateTime.parse`. Whether it should be backported or not is the question - I think it's probably safe on balance because it'll only affect values that were previously blocked.

**Comment by trcarden** [comment]
@arnvald please add tests and open a PR. I probably won't be able to get to this right away

**Comment by arnvald** [comment]
@pixeltrix should it be just `Time.iso8601(data)` or `Time.iso8601(data).to_datetime`? I'm not sure how important it is to return `DateTime` class - if it's necessary then the 2nd snippet will solve it.

**Comment by pixeltrix** [comment]
@arnvald I would think that returning a `Time` is preferable to a `DateTime` since IS0 8601 is all about a proleptic gregorian calendar.

**Comment by arnvald** [comment]
@pixeltrix using `Time.iso8601` introduces another problem, this test starts failing: https://github.com/rails/rails/blob/master/activesupport/test/json/decoding_test.rb#L22

Here's comparison
```ruby
[3] pry(main)> DateTime.parse("2015-10-20") # old version
=> Tue, 20 Oct 2015 00:00:00 +0000
[4] pry(main)> Time.iso8601("2015-10-20") # new version
ArgumentError: invalid date: "2015-10-20"
from /Users/Gregor/.rubies/2.2/lib/ruby/2.2.0/time.rb:596:in `xmlschema'
```

Alternatively, `Time.parse` can be used. It handles all the cases, the only difference is that it uses local timezone for `YYY-MM-DD` date format:
```ruby
[30] pry(main)> DateTime.parse "2015-10-20" # old version
=> Tue, 20 Oct 2015 00:00:00 +0000
[31] pry(main)> Time.parse "2015-10-20" # new version
=> 2015-10-20 00:00:00 +0800
```

**Comment by chancancode** [comment]
—

**Comment by pixeltrix** [comment]
There's a `Date.iso8601` which handles that correctly. Perhaps `YYYY-MM-DD` should return `Date` objects?

**Comment by arnvald** [comment]
@pixeltrix @chancancode I've been working on this issue, and the last thing that's causing me a problem is that when I call `Time.parse` (`Time.iso8601` failes to parse `YYYY-MM-DD hh:mm:ss` case - it requires `T` instead of space), system local time is used, no matter what timezone is defined in the application. I think an application timezone is expected, am I right?

An alternative is to use `Time.zone.parse` - it handles all the cases correctly, but it has performance hit (it's about 1.6x slower than `Time.parse`).

You can see the changes here: https://github.com/arnvald/rails/commit/35cd28ebffbd3cbae4e2d8ac0f69058a0173c4e9

If you think we should always use system local time, or that there's some better solution, I'll update it.

**Comment by maclover7** [comment]
@arnvald Can you open up a PR with your changes, please?

**Comment by chancancode** [comment]
Thanks everyone, sorry this took so long! It's fixed in #23011/master now. I'll check if we can backport this into the release branches.
